### PR TITLE
Fixes #20998 - Allow filtering by action column

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: db/migrate/20171026082635_add_task_action.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class AddTaskAction < ActiveRecord::Migration
+  def up
+    add_column :foreman_tasks_tasks, :action, :string
+
+    ::ForemanTasks::Task.where(:action => nil).find_each do |task|
+      task.action = task.format_input
```
Comment (iNecas) [inline db/migrate/20171026082635_add_task_action.rb]: This can be pretty expensive operation, that would slow down the whole migration a lot. I wonder, if we should do it here? Perhaps a rake task to calculate this, if anyone wanted that, but in general, I'm not sure many people would want that for older tasks: if the did, they could run the rake task.

Comment (adamruzicka) [reply db/migrate/20171026082635_add_task_action.rb]: The way the UI is now, without actually computing the changes all the old tasks would have blank action column in the tasks index. We could fallback to generating this stuff on demand when the user displays the task, but it might break the search.

Comment (iNecas) [reply db/migrate/20171026082635_add_task_action.rb]: :+1: for fallback

**Comment by adamruzicka** [comment]
The data is now not migrated when running the database migration, there is a rake task for doing that. When the user accesses the WebUI, we check for the data in DB and fall back to generating it on demand. This however breaks the search because only the in-db data is searchable. For this reason a `foreman_tasks:generate_task_actions`, which populates the action column is included. This is not needed for tasks started after this db migration, because they regenerate the column as they change states

File: lib/foreman_tasks/tasks/generate_task_actions.rake [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# Run "foreman-rake foreman_tasks:generate_task_actions" to generate missing values.
+
+namespace :foreman_tasks do
+  desc 'Generate missing values for action column in foreman_tasks_tasks table.'
+  task :generate_task_actions => :environment do
+    scope = ::ForemanTasks::Task.where(:action => nil)
+
+    puts "Generating action for #{scope.count} tasks."
+    scope.find_each do |task|
+      task.action = task.format_input
```
Comment (iNecas) [inline lib/foreman_tasks/tasks/generate_task_actions.rake]: Some info about the progress would be nice. Let's say printing `done/total', every 100 records or something like that, to give the user some idea on when to expect it will finish.

File: app/models/foreman_tasks/task.rb [hunk 3]
```diff
@@ -36,6 +36,7 @@ def check_permissions_after_save
     end
 
     scoped_search :on => :id, :complete_value => false
+    scoped_search :on => :action, :complete_value => false
```
Comment (ShimShtein) [inline app/models/foreman_tasks/task.rb]: I see we are adding yet another implicit field to search. IMHO it would be better to be more explicit, it will make the query simpler and run faster. You can do it easily by adding `:only_explicit => true`

File: app/models/foreman_tasks/task/dynflow_task.rb [hunk 4]
```diff
@@ -19,6 +19,7 @@ def update_from_dynflow(data)
                                 end
                               end
       self.label          ||= main_action && main_action.class.name
+      self.action = format_input(true)
```
Comment (ShimShtein) [inline app/models/foreman_tasks/task/dynflow_task.rb]: Wouldn't it be better to put action calculation to `:after_validate` callback? This way we will be sure it's updated each time we save the model.

File: app/models/foreman_tasks/task.rb (hunk 3 above)
Comment (iNecas) [reply app/models/foreman_tasks/task.rb]: For this particular, I would probably leave this in the implicit search TBH

File: app/models/foreman_tasks/task/dynflow_task.rb (hunk 4 above)
Comment (iNecas) [reply app/models/foreman_tasks/task/dynflow_task.rb]: Would work for me. Also: would it work if we did `self.action ||= `, to save some calcutlation time on each save?

Comment (ShimShtein) [reply app/models/foreman_tasks/task/dynflow_task.rb]: > would it work if we did self.action ||=, to save some calcutlation time on each save?

It depends if `:name` and `:input` could be changed. For clarity I would suggest using condition on the callback:
``` ruby
after_validation :set_action_field :if => Proc.new { |task| task.name_changed? || task.input_changed? }
```

File: db/migrate/20171026082635_add_task_action.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class AddTaskAction < ActiveRecord::Migration
```
Comment (ShimShtein) [inline db/migrate/20171026082635_add_task_action.rb]: Could you please generate the migration using our plugin specific task? See our [wiki](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Add-new-migration) for reference.

TL;DR: Please rename the migration to `20171026082635_add_task_action.foreman-tasks.rb`. It will be the first step to dynamic add and remove plugins.

File: lib/foreman_tasks/tasks/generate_task_actions.rake [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        @processed = 0
+        puts message % { :count => count } unless message.nil?
+      end
+
+      def progress(count)
+        @processed += count
+      end
+
+      def report
+        puts "Processed #{@processed}/#{@count} tasks"
```
Comment (ShimShtein) [inline lib/foreman_tasks/tasks/generate_task_actions.rake]: This needs i18n IMHO...

File: lib/foreman_tasks/tasks/generate_task_actions.rake [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#
+# generate_task_actions.rake is a data migration task to properly fill the missing
+# values into the foreman_tasks_tasks table.
+#
+# Run "foreman-rake foreman_tasks:generate_task_actions" to generate missing values.
+
+namespace :foreman_tasks do
+  desc 'Generate missing values for action column in foreman_tasks_tasks table.'
```
Comment (ShimShtein) [inline lib/foreman_tasks/tasks/generate_task_actions.rake]: This needs i18n I believe

File: app/helpers/foreman_tasks/tasks_helper.rb [hunk 8]
```diff
@@ -2,16 +2,7 @@ module ForemanTasks
...
-      parts << task.get_humanized(:name) if include_action
-      parts << Array(task.get_humanized(:input)).map do |part|
-        if part.is_a? Array
-          part[1][:text]
-        else
-          part.to_s
-        end
-      end.join('; ')
-      parts.join(' ')
+      task.action || task.format_input(include_action)
```
Comment (ShimShtein) [inline app/helpers/foreman_tasks/tasks_helper.rb]: Helper shouldn't know how `action` field is generated. I would prefer leaving this calculation to the model:
``` ruby
# In task.rb:
class Task < ActiveRecord::Base
  def action
    super || format_input
  end
end
```

File: app/models/foreman_tasks/task.rb [hunk 9]
```diff
@@ -202,6 +203,19 @@ def sub_tasks_counts
       result.symbolize_keys
     end
 
+    def format_input(include_action = false)
```
Comment (ShimShtein) [inline app/models/foreman_tasks/task.rb]: I think we should get rid of `include_action` parameter. It will cause inconsistent output in the `format_task_input` helper, depending on the fact was the task migrated or created by `dynflow_task`.

File: lib/foreman_tasks/tasks/generate_task_actions.rake [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      end
+
+      def report
+        puts "Processed #{@processed}/#{@count} tasks"
+      end
+    end
+
+    scope = ::ForemanTasks::Task.where(:action => nil)
+    count = scope.count
+    reporter = ProgressReporter.new count, 'Generating action for %{count} tasks.'
```
Comment (ShimShtein) [inline lib/foreman_tasks/tasks/generate_task_actions.rake]: This needs i18n too...

File: app/models/foreman_tasks/task/dynflow_task.rb (hunk 4 above)
Comment (adamruzicka) [reply app/models/foreman_tasks/task/dynflow_task.rb]: The thing here is that `name` and `input` are not attributes of the model so we're not getting all the `*_changed?` goodies from rails. Not sure about `name`, but `input` can definitely change during the run of the task so `||=` is probably not a good idea. Any other idea than implementing `*_changed?` ourselves?

Comment (ShimShtein) [reply app/models/foreman_tasks/task/dynflow_task.rb]: If we can't track changes to `name` and `input`, we will not be able to avoid recalculation. We don't have to write field tracking code from scratch, since we already have it in our models from [ActiveMode::Dirty](http://api.rubyonrails.org/classes/ActiveModel/Dirty.html). Anyway, the field will be omitted from SQL UPDATE statement, if the value remains the same (tested it on another ActiveRecord). I don't think the calculation is that expensive that we should try and optimize it. IMHO, we can set the field each save without any worries.

File: lib/foreman_tasks/tasks/generate_task_actions.rake (hunk 7 above)
Comment (adamruzicka) [reply lib/foreman_tasks/tasks/generate_task_actions.rake]: There is no i18n context defined when the rake tasks are being defined. Added translations to all the other places

**Comment by ShimShtein** [comment]
@adamruzicka, This needs a rebase. Also what about i18n of the rake task description?

**Comment by adamruzicka** [comment]
Rebase is in the works, the rake task's description is being defined outside any i18n context so it probably can't be translated easily. Also none of rake tasks in foreman-tasks has translated description

File: db/migrate/20171026082635_add_task_action.foreman_tasks.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class AddTaskAction < ActiveRecord::Migration
```
Comment (iNecas) [inline db/migrate/20171026082635_add_task_action.foreman_tasks.rb]: Needs to be updated to `Rails 5` syntax.

File: lib/foreman_tasks/tasks/generate_task_actions.rake [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      def progress(count)
+        @processed += count
+      end
+
+      def report
+        puts _('Processed %{processed}/%{count} tasks') % { :processed => @processed, :count => @count }
+      end
+    end
+
+    scope = ::ForemanTasks::Task.where(:action => nil)
```
Comment (iNecas) [inline lib/foreman_tasks/tasks/generate_task_actions.rake]: What about ordering by `started_at desc`? This way, when somebody runs this and cuts it in the middle (because she doesn't want to wait until the script to finish), she still can search in the most recent/relevant tasks.

**Comment by iNecas** [comment]
Thanks @adamruzicka
